flink/v1.12/flink/src/main/java/com/netease/arctic/flink/read/source/log/LogKafkaPartitionSplitReader.java

```diff
@@ -0,0 +1,369 @@
+        // Stop consuming from this partition if the offsets has reached the stopping
+        // offset.
+        // Note that there are two cases, either case finishes a split:
+        // 1. After processing a record with offset of "stoppingOffset - 1". The split
+        // reader
+        //    should not continue fetching because the record with stoppingOffset may not
+        // exist.
+        // 2. Before processing a record whose offset is greater than or equals to the
+        // stopping
+        //    offset. This should only happens when case 1 was not met due to log compaction
+        // or
+        //    log retention.
+        // Case 2 is handled here. Case 1 is handled after the record is processed.
+        if (consumerRecord.offset() >= stoppingOffset) {
+          finishSplitAtRecord(
+              tp,
+              stoppingOffset,
+              consumerRecord.offset(),
+              finishedPartitions,
+              recordsBySplits);
+          break;
+        }
+
+        byte[] value = consumerRecord.value();
+        boolean magicFormat = checkMagicNum(value);
+        if (!magicFormat) {
+          throw new UnsupportedOperationException(
+              "Can't deserialize arctic log queue message due to it does not contain magic number.");
+        }
+
+        LogData<RowData> logData = logDataJsonDeserialization.deserialize(value);
+        if (!logData.getFlip() && filterByRowKind(logData.getActualValue())) {
+          LOG.info(
+              "filter the rowData, because of logConsumerAppendOnly is true, and rowData={}.",
+              logData.getActualValue());
+          continue;
+        }
+
+        // Finish the split because there might not be any message after this point.
+        // Keep polling
+        // will just block forever.
+        if (consumerRecord.offset() == stoppingOffset - 1) {
+          finishSplitAtRecord(
+              tp,
+              stoppingOffset,
+              consumerRecord.offset(),
+              finishedPartitions,
+              recordsBySplits);
+        }
+
+        final long currentOffset = consumerRecord.offset();
+
+        if (logData.getFlip()) {
+          if (logRetractionEnable) {
+            logReadHelper.startRetracting(tp, logData.getUpstreamId(), logData.getEpicNo(),
+                currentOffset + 1);
+            break;
+          } else {
+            continue;
+          }
+        }
+
+        if (logRetractionEnable) {
+          logReadHelper.initialEpicStartOffsetIfEmpty(tp, logData.getUpstreamId(), logData.getEpicNo(), currentOffset);
+        }
+        recordsForSplit.add(LogRecordWithRetractInfo.of(consumerRecord, logData));
+      }
+    }
+    // Unassign the partitions that has finished.
+    if (!finishedPartitions.isEmpty()) {
+      unassignPartitions(finishedPartitions);
+    }
+  }
+
+  /**
+   * read reversely in retracting mode
+   */
+  private void readReversely(Set<TopicPartition> retractTps,
```

Comment:
Could add more tests for it?
The mock data should contain four rowKinds in the UT which is better.